### PR TITLE
fix(datamodel): sort parts by numerical price in part selector and BOM lists (fixes #846)

### LINKED ISSUE #846: Incorrect sort by price
**Describe the bug**
Sorting by price is incorrect. For example, $0.05 should not sort between $0.40 and $0.60.

**To Reproduce**
1. Click "JLCPCB Tools" in toolbar
2. Select a part
3. Click "Assign LCSC Number"
4. Type "USB Type-C" into the field and press Enter
5. Click the "Price" column header to sort in ascending order
6. Drag the "Price" column wider

**Expected behavior**
Prices should be in ascending order.

**Actual behavior**

Image

# KiCad Version
```
Application: KiCad PCB Editor x86_64 on x86_64

Version: 10.0.6, release build

Libraries:
	wxWidgets 3.2.8 
	FreeType 2.14.3
	HarfBuzz 14.3.0
	FontConfig 2.18.2
	libcurl/8.7.1 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.12 nghttp2/1.64.0

Platform: macOS Sequoia Version 15.7.3 (Build 24G419), 64 bit, Little endian, wxMac
OpenGL: ATI Technologies Inc., AMD Radeon R9 M370X OpenGL Engine, 2.1 ATI-4.8.101

Build Info:
	Date: Aug 28 2026 15:31:44
	wxWidgets: 3.2.8 (wchar_t,wx containers)
	Boost: 1.90.0
	OCC: 7.9.3
	Curl: 8.7.1
	ngspice: 45.2
	Compiler: Clang 16.0.0 with C++ ABI 1002
	KICAD_IPC_API=ON
	KICAD_USE_PCH=OFF
```

### PR DESCRIPTION
**Pull request by z2amiller**
- Parse price strings, quantity tiers, and raw numbers into float values in bom_estimation.pricing.parse_price
- Provide price_sort_key with ascending/descending handling that sorts errors and blanks to the end
- Add SQLite pricesort collation for library.search queries ordering by Price
- Connect price_column sorting in _StockDataModel, PartSelectorDataModel, and PartListDataModel
- Collapse redundant tier parsing in PartSelectorDialog.get_price to delegate to get_unit_price
- Add unit tests in tests/test_price_sorting.py verifying issue #846 fix and boundary conditions

fixes: #846

### CODE REVIEW

File: library.py [hunk 1]
```diff
@@ -461,13 +462,15 @@ def search(self, parameters):
                 query += " AND "
             query += " AND ".join(query_chunks)
 
-        query += f' ORDER BY "{self.order_by}" COLLATE naturalsort {self.order_dir}'
+        collation = "pricesort" if self.order_by == "Price" else "naturalsort"
+        query += f' ORDER BY "{self.order_by}" COLLATE {collation} {self.order_dir}'
```
[the remark this answers is not in the document]
Comment (z2amiller) [reply library.py]: Good catch. We updated `price_sort_collation` to be direction-aware and registered `pricesort_asc` and `pricesort_desc` collations in SQLite. In descending searches, queries now evaluate `ORDER BY "Price" COLLATE pricesort_desc ASC LIMIT 1000` so invalid and blank prices remain at the end in both directions and never displace valid parts under `LIMIT`.

File: library.py [hunk 2]
```diff
@@ -461,13 +462,28 @@ def search(self, parameters):
...
             query += " AND ".join(query_chunks)
 
-        query += f' ORDER BY "{self.order_by}" COLLATE naturalsort {self.order_dir}'
+        if self.order_by == "Price":
+            collation = (
+                "pricesort_desc"
+                if str(self.order_dir).upper() == "DESC"
+                else "pricesort_asc"
+            )
+            query += f' ORDER BY "{self.order_by}" COLLATE {collation} ASC'
```
[the remark this answers is not in the document]
Comment (z2amiller) [reply library.py:472]: Implemented in commit 0b79e01. We updated `parse_price`, `price_sort_key`, and `price_sort_collation` to accept an optional `quantity` argument (defaulting to 1). `partselector.py` now passes `quantity=len(self.parts)` through `parameters`, and `library.search` binds that quantity to `pricesort_asc` and `pricesort_desc` so the SQLite database query orders parts using the exact quantity-tier prices displayed to the user.

**Review by chmorgan — APPROVED** [review]
Reviewed `104234a`. The numeric comparisons address #846, and the fixes for descending invalid values and selected-quantity tiers look correct.

Validation: 248 relevant tests passed, changed-file Ruff lint and format checks passed, and additional real SQLite searches verified both directions across quantity boundaries and the 1,000-result limit. Native macOS UI verification was not completed because wx could not initialize with screen access in the test environment.

Nonblocking follow-up: `OnSortPartList` is already unbound in the base branch, so Price header clicks still sort only the loaded results; the new SQL ordering is not reached through that UI action. This is an existing limitation, not a regression introduced here.
